api/handlers.go

```diff
@@ -268,79 +269,121 @@ func (api *API) userPOST(w http.ResponseWriter, req *http.Request, _ httprouter.
 		api.WriteError(w, err, http.StatusBadRequest)
 		return
 	}
-	if payload.StripeID == "" {
-		err = errors.AddContext(err, "empty stripe id")
-		api.WriteError(w, err, http.StatusBadRequest)
+	if payload.Email == "" && payload.StripeID == "" {
+		// The payload is empty, nothing to do.
+		api.WriteSuccess(w)
 		return
 	}
-	// Check if this user already has this ID assigned to them.
-	if payload.StripeID == u.StripeID {
-		// Nothing to do.
-		api.WriteJSON(w, u)
+
+	// Fetch the user from the DB.
+	u, err := api.staticDB.UserBySub(req.Context(), sub, false)
+	if errors.Contains(err, database.ErrUserNotFound) {
+		api.WriteError(w, err, http.StatusNotFound)
 		return
 	}
-	// Check if a user already has this customer id.
-	eu, err := api.staticDB.UserByStripeID(req.Context(), payload.StripeID)
-	if err != nil && err != database.ErrUserNotFound {
+	if err != nil {
 		api.WriteError(w, err, http.StatusInternalServerError)
 		return
 	}
-	if err == nil && eu.ID.Hex() != u.ID.Hex() {
-		err = errors.New("this stripe customer id belongs to another user")
-		api.WriteError(w, err, http.StatusBadRequest)
-		return
+
+	if payload.StripeID != "" {
+		// Check if this user already has a Stripe customer ID.
+		if u.StripeID != "" {
+			err = errors.New("this user already has a Stripe customer id")
+			api.WriteError(w, err, http.StatusUnprocessableEntity)
+			return
+		}
+		// Verify that no other user owns this StripeID.
+		su, err := api.staticDB.UserByStripeID(req.Context(), payload.StripeID)
+		if err != nil && !errors.Contains(err, database.ErrUserNotFound) {
+			api.WriteError(w, err, http.StatusInternalServerError)
+			return
+		}
+		if err == nil && su.Sub != sub {
+			err = errors.New("this stripe customer id belongs to another user")
+			api.WriteError(w, err, http.StatusBadRequest)
+			return
+		}
+		// Set the StripeID.
+		u.StripeID = payload.StripeID
 	}
-	// Check if this user already has a Stripe customer ID.
-	if u.StripeID != "" {
-		err = errors.New("this user already has a Stripe customer id")
-		api.WriteError(w, err, http.StatusUnprocessableEntity)
-		return
+
+	var changedEmail bool
+	if payload.Email != "" {
+		// Validate the new email.
+		a, err := mail.ParseAddress(payload.Email)
+		if err != nil {
+			api.WriteError(w, errors.AddContext(err, "invalid email address"), http.StatusBadRequest)
+			return
+		}
+		// Strip any names from the email and leave just the address.
+		payload.Email = a.Address
+		// Check if another user already has this email address.
+		eu, err := api.staticDB.UserByEmail(req.Context(), payload.Email, false)
+		if err != nil && !errors.Contains(err, database.ErrUserNotFound) {
+			api.WriteError(w, err, http.StatusInternalServerError)
+			return
+		}
+		if err == nil && eu.Sub != sub {
+			err = errors.New("this email is already in use")
+			api.WriteError(w, err, http.StatusBadRequest)
+			return
+		}
+		// Set the new email and set it up for a confirmation.
+		u.Email = payload.Email
```

Comment:
Yes, because we cannot be sure that they didn't mess up their email or that they own that email. If we forgo confirmation here, we might as well not do it during registration.

One valid concern here is what to do if a user changes their email to a misspelled one (so, they do not have access) and then forgets their password. They will want to recover the account but it would be already pointing to the unconfirmed email, that they don't have access to anyway. This situation is similar to a user registering with an email they don't control and then forgetting their password. Similar but not the same, as in the "change" case we can still allow them to use the previous email for recovery until they confirm the new one. I'd be happy to do that in a F/U.